File: lib/models/customer.dart
class Customer {
  final int? id;
  final String customerName;
  final String? phoneNumber;
  final String? email;
  final String? address;
  final String? gstNumber;
  final String? notes;
  final double? totalPurchases;
  final int? totalInvoices;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Customer({
    this.id,
    required this.customerName,
    this.phoneNumber,
    this.email,
    this.address,
    this.gstNumber,
    this.notes,
    this.totalPurchases,
    this.totalInvoices,
    this.createdAt,
    this.updatedAt,
  });

  factory Customer.fromJson(Map<String, dynamic> json) {
    return Customer(
      id: json['id'],
      customerName: json['customerName'] ?? '',
      phoneNumber: json['phoneNumber'],
      email: json['email'],
      address: json['address'],
      gstNumber: json['gstNumber'],
      notes: json['notes'],
      totalPurchases: json['totalPurchases']?.toDouble(),
      totalInvoices: json['totalInvoices'],
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() => {
    if (id != null) 'id': id,
    'customerName': customerName,
    if (phoneNumber != null) 'phoneNumber': phoneNumber,
    if (email != null) 'email': email,
    if (address != null) 'address': address,
    if (gstNumber != null) 'gstNumber': gstNumber,
    if (notes != null) 'notes': notes,
  };
}

File: backend/src/main/java/com/billingapp/entity/Customer.java
package com.billingapp.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Size;
import lombok.*;

import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "customers")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Customer extends BaseEntity {

    @NotBlank
    @Size(max = 100)
    @Column(name = "customer_name")
    private String customerName;

    @Column(name = "phone_number")
    private String phoneNumber;

    @Column(name = "email")
    private String email;

    @Column(name = "address", columnDefinition = "TEXT")
    private String address;

    @Column(name = "gst_number")
    private String gstNumber;

    @Column(name = "notes", columnDefinition = "TEXT")
    private String notes;

    @Column(name = "total_purchases", precision = 12, scale = 2)
    private java.math.BigDecimal totalPurchases = java.math.BigDecimal.ZERO;

    @Column(name = "total_invoices")
    private Integer totalInvoices = 0;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shop_id", nullable = false)
    private Shop shop;

    @OneToMany(mappedBy = "customer", cascade = CascadeType.ALL)
    @Builder.Default
    private List<Invoice> invoices = new ArrayList<>();
}
